File: app/src/main/java/one/mixin/android/widget/ButtonGroupLayout.java
package one.mixin.android.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

public class ButtonGroupLayout extends ViewGroup {
    private int lineSpacing;

    public ButtonGroupLayout(Context context) {
        this(context, null);
    }

    public ButtonGroupLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ButtonGroupLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        lineSpacing = 0;
    }

    public void setLineSpacing(int lineSpacing) {
        this.lineSpacing = lineSpacing;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);

        int width = 0;
        int height = getPaddingTop() + getPaddingBottom();

        int childCount = getChildCount();
        int maxChildHeight = 0;
        int lineWidth = getPaddingLeft() + getPaddingRight();

        for (int i = 0; i < childCount; i++) {
            View child = getChildAt(i);
            measureChild(child, widthMeasureSpec, heightMeasureSpec);
            maxChildHeight = Math.max(maxChildHeight, child.getMeasuredHeight());
            lineWidth += child.getMeasuredWidth() + lineSpacing;

            if ((i + 1) % 3 == 0 || i == childCount - 1) {
                height += maxChildHeight;
                lineWidth = getPaddingLeft() + getPaddingRight();
                maxChildHeight = 0;
            }
        }

        setMeasuredDimension(
                widthMode == MeasureSpec.EXACTLY ? widthSize : width,
                heightMode == MeasureSpec.EXACTLY ? heightSize : height
        );
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int width = getWidth() - getPaddingLeft() - getPaddingRight();
        int childCount = getChildCount();
        int lineWidth = getPaddingLeft();
        int top = getPaddingTop();
        int maxChildHeight = 0;

        for (int i = 0; i < childCount; i++) {
            View child = getChildAt(i);
            int childrenInRow = Math.min(3, childCount - i);

            if (childrenInRow == 1) {
                int childWidth = width;
                child.measure(
                        MeasureSpec.makeMeasureSpec(childWidth, MeasureSpec.EXACTLY),
                        MeasureSpec.makeMeasureSpec(child.getMeasuredHeight(), MeasureSpec.EXACTLY)
                );
                child.layout(lineWidth, top, lineWidth + childWidth, top + child.getMeasuredHeight());
            } else if (childrenInRow == 2) {
                int childWidth = (width - lineSpacing) / 2;
                for (int j = 0; j < childrenInRow; j++) {
                    View subChild = getChildAt(i + j);
                    subChild.measure(
                            MeasureSpec.makeMeasureSpec(childWidth, MeasureSpec.EXACTLY),
                            MeasureSpec.makeMeasureSpec(subChild.getMeasuredHeight(), MeasureSpec.EXACTLY)
                    );
                    subChild.layout(lineWidth, top, lineWidth + childWidth, top + subChild.getMeasuredHeight());
                    lineWidth += childWidth + lineSpacing;
                }
            } else {
                int childWidth = (width - 2 * lineSpacing) / 3;
                for (int j = 0; j < childrenInRow; j++) {
                    View subChild = getChildAt(i + j);
                    subChild.measure(
                            MeasureSpec.makeMeasureSpec(childWidth, MeasureSpec.EXACTLY),
                            MeasureSpec.makeMeasureSpec(subChild.getMeasuredHeight(), MeasureSpec.EXACTLY)
                    );
                    subChild.layout(lineWidth, top, lineWidth + childWidth, top + subChild.getMeasuredHeight());
                    lineWidth += childWidth + lineSpacing;
                }
            }

            maxChildHeight = Math.max(maxChildHeight, child.getMeasuredHeight());
            top += maxChildHeight;
            lineWidth = getPaddingLeft();
            maxChildHeight = 0;
            i += childrenInRow - 1;
        }
    }
}

File: app/src/main/java/one/mixin/android/widget/ButtonGroupLayout.kt
package one.mixin.android.widget

import android.content.Context
import android.util.AttributeSet
import android.view.ViewGroup

class ButtonGroupLayout(context: Context, attrs: AttributeSet?) : ViewGroup(context, attrs) {

    private var horizontalSpacing = 0
    private var verticalSpacing = 0

    fun setLineSpacing(lineSpacing: Int) {
        horizontalSpacing = lineSpacing
        verticalSpacing = lineSpacing
    }

    private val layoutGird = mutableListOf<Int>()

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        var totalHeight = 0
        layoutGird.clear()

        val miniWidth = (widthSize - paddingStart - paddingEnd - horizontalSpacing * 2) / 3
        val middleWidth = (widthSize - paddingStart - paddingEnd - horizontalSpacing) / 2
        val maxWidth = widthSize - paddingStart - paddingEnd

        for (i in 0 until childCount) {
            val child = getChildAt(i)
            measureChild(child, MeasureSpec.makeMeasureSpec(maxWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
        }

        var i = 0
        while (i < childCount) {
            if (i + 2 < childCount) {
                val firstView = getChildAt(i)
                val secondView = getChildAt(i + 1)
                val thirdView = getChildAt(i + 2)
                if (firstView.measuredWidth + secondView.measuredWidth + thirdView.measuredWidth + horizontalSpacing * 2 <= maxWidth && firstView.measuredWidth<= miniWidth && secondView.measuredWidth <= miniWidth && thirdView.measuredWidth <= miniWidth) {
                    totalHeight += listOf(firstView.measuredHeight, secondView.measuredHeight, thirdView.measuredHeight).max() + verticalSpacing
                    layoutGird.add(1)
                    layoutGird.add(1)
                    layoutGird.add(1)
                    firstView.measure(MeasureSpec.makeMeasureSpec(miniWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    secondView.measure(MeasureSpec.makeMeasureSpec(miniWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    thirdView.measure(MeasureSpec.makeMeasureSpec(miniWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    i += 3
                } else if (firstView.measuredWidth + secondView.measuredWidth + horizontalSpacing <= maxWidth && firstView.measuredWidth<= middleWidth && secondView.measuredWidth <= middleWidth) {
                    totalHeight += listOf(firstView.measuredHeight, secondView.measuredHeight).max() + verticalSpacing
                    firstView.measure(MeasureSpec.makeMeasureSpec(middleWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    secondView.measure(MeasureSpec.makeMeasureSpec(middleWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    layoutGird.add(2)
                    layoutGird.add(2)
                    i += 2
                } else {
                    totalHeight += firstView.measuredHeight + verticalSpacing
                    firstView.measure(MeasureSpec.makeMeasureSpec(maxWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    layoutGird.add(3)
                    i += 1
                }
            } else if (i + 1 < childCount) {
                val firstView = getChildAt(i)
                val secondView = getChildAt(i + 1)
                if (firstView.measuredWidth + secondView.measuredWidth + horizontalSpacing <= maxWidth &&  firstView.measuredWidth<= middleWidth && secondView.measuredWidth <= middleWidth) {
                    totalHeight += listOf(firstView.measuredHeight, secondView.measuredHeight).max() + verticalSpacing
                    firstView.measure(MeasureSpec.makeMeasureSpec(middleWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    secondView.measure(MeasureSpec.makeMeasureSpec(middleWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    layoutGird.add(2)
                    layoutGird.add(2)
                    i += 2
                } else {
                    totalHeight += firstView.measuredHeight + verticalSpacing
                    firstView.measure(MeasureSpec.makeMeasureSpec(maxWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                    layoutGird.add(3)
                    i += 1
                }
            } else if (i < childCount) {
                val firstView = getChildAt(i)
                totalHeight += firstView.measuredHeight + verticalSpacing
                firstView.measure(MeasureSpec.makeMeasureSpec(maxWidth, MeasureSpec.EXACTLY), heightMeasureSpec)
                layoutGird.add(3)
                i += 1
            } else {
                break
            }
        }
        totalHeight -= verticalSpacing

        setMeasuredDimension(widthSize, totalHeight + paddingTop + paddingBottom)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val miniWidth = (measuredWidth - paddingStart - paddingEnd - horizontalSpacing * 2) / 3
        val middleWidth = (measuredWidth - paddingStart - paddingEnd - horizontalSpacing) / 2
        val maxWidth = measuredWidth - paddingStart - paddingEnd
        var curTop = paddingTop
        var curLeft = paddingStart
        var i = 0
        while (i < childCount)
            if (layoutGird[i] == 1) {
                val firstView = getChildAt(i)
                val secondView = getChildAt(i + 1)
                val thirdView = getChildAt(i + 2)
                firstView.layout(curLeft, curTop, curLeft + miniWidth, curTop + firstView.measuredHeight)
                curLeft += miniWidth + horizontalSpacing
                secondView.layout(curLeft, curTop, curLeft + miniWidth, curTop + secondView.measuredHeight)
                curLeft += miniWidth + horizontalSpacing
                thirdView.layout(curLeft, curTop, curLeft + miniWidth, curTop + thirdView.measuredHeight)
                curLeft = paddingStart
                curTop += listOf(firstView.measuredHeight, secondView.measuredHeight, thirdView.measuredHeight).max() + verticalSpacing
                i += 3
            } else if (layoutGird[i] == 2) {
                val firstView = getChildAt(i)
                val secondView = getChildAt(i + 1)
                firstView.layout(curLeft, curTop, curLeft + middleWidth, curTop + firstView.measuredHeight)
                curLeft += middleWidth + horizontalSpacing
                secondView.layout(curLeft, curTop, curLeft + middleWidth, curTop + secondView.measuredHeight)
                curLeft = paddingStart
                curTop += listOf(firstView.measuredHeight, secondView.measuredHeight).max() + verticalSpacing
                i += 2
            } else {
                val firstView = getChildAt(i)
                firstView.layout(curLeft, curTop, curLeft + maxWidth, curTop + firstView.measuredHeight)
                curLeft = paddingStart
                curTop += firstView.measuredHeight + verticalSpacing
                i += 1
            }
    }
}
